From: Amit Shah <amit.shah@redhat.com>
Date: Tue, 31 Aug 2010 05:35:05 -0400
Subject: [virtio] fix balloon without VIRTIO_BALLOON_F_STATS_VQ
Message-id: <2a8d9b9bc0b81bc57f4b055117d390b5788049a3.1283231512.git.amit.shah@redhat.com>
Patchwork-id: 27956
O-Subject: [RHEL5.6 PATCH 3/3] virtio: fix balloon without
	VIRTIO_BALLOON_F_STATS_VQ
Bugzilla: 601692
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Rusty Russell <rusty@rustcorp.com.au>

When running under qemu-kvm-0.11.0:

	BUG: unable to handle kernel paging request at 56e58955
	...
	Process vballoon (pid: 1297, ti=c7976000 task=c70a6ca0 task.ti=c7
	...
	Call Trace:
	 [<c88253a3>] ? balloon+0x1b3/0x440 [virtio_balloon]
	 [<c041c2d7>] ? schedule+0x327/0x9d0
	 [<c88251f0>] ? balloon+0x0/0x440 [virtio_balloon]
	 [<c014a2d4>] ? kthread+0x74/0x80
	 [<c014a260>] ? kthread+0x0/0x80
	 [<c0103b36>] ? kernel_thread_helper+0x6/0x30

need_stats_update should be zero-initialized.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
Acked-by: Adam Litke <agl@us.ibm.com>
(cherry picked from commit 169c246a30808588436794e96a97c61a01af9bed)

Signed-off-by: Amit Shah <amit.shah@redhat.com>

diff --git a/drivers/virtio/virtio_balloon.c b/drivers/virtio/virtio_balloon.c
index 3341851..268a1c4 100644
--- a/drivers/virtio/virtio_balloon.c
+++ b/drivers/virtio/virtio_balloon.c
@@ -295,6 +295,7 @@ static int virtballoon_probe(struct virtio_device *vdev)
 	vb->num_pages = 0;
 	init_waitqueue_head(&vb->config_change);
 	vb->vdev = vdev;
+	vb->need_stats_update = 0;
 
 	/* We expect two virtqueues: inflate and deflate,
 	 * and optionally stat. */
